crates/lang/macro/src/lib.rs

```diff
@@ -18,20 +18,600 @@ mod trait_def;
 }
 
+/// Marks trait definitions to ink! as special ink! trait definitions.
+///
+/// There are some restrictions that apply to ink! trait definitions that
+/// this macro checks. Also ink! trait definitions are required to have specialized
+/// structure so that the main [`#[ink::contract]`](`macro@crate::contract`) macro can
+/// properly generate code for its implementations.
+///
+/// # Example: Definition
+///
+/// ```
+/// use ink_lang as ink;
+/// # type Balance = <ink_env::DefaultEnvTypes as ink_env::EnvTypes>::Balance;
+///
+/// #[ink::trait_definition]
+/// pub trait Erc20 {
+///     /// Constructors a new ERC-20 compliant smart contract using the initial supply.
+///     #[ink(constructor)]
+///     fn new(initial_supply: Balance) -> Self;
+///
+///     /// Returns the total supply of the ERC-20 smart contract.
+///     #[ink(message)]
+///     fn total_supply(&self) -> Balance;
+///
+///     // etc.
+/// }
+/// ```
+///
+/// # Example: Implementation
+///
+/// Given the above trait definition you can implement it as shown below:
+///
+/// ```
+/// # use ink_lang as ink;
+/// #
+/// #[ink::contract]
+/// mod base_erc20 {
+/// #    // We somehow cannot put the trait in the doc-test crate root due to bugs.
+/// #    #[ink_lang::trait_definition]
+/// #    pub trait Erc20 {
+/// #        /// Constructors a new ERC-20 compliant smart contract using the initial supply.
+/// #        #[ink(constructor)]
+/// #        fn new(initial_supply: Balance) -> Self;
+/// #
+/// #        /// Returns the total supply of the ERC-20 smart contract.
+/// #        #[ink(message)]
+/// #        fn total_supply(&self) -> Balance;
+/// #    }
+/// #
+///     #[ink(storage)]
+///     pub struct BaseErc20 {
+///         total_supply: Balance,
+///         // etc ..
+///     }
+///
+///     impl Erc20 for BaseErc20 {
+///         #[ink(constructor)]
+///         fn new(initial_supply: Balance) -> Self {
+///             Self { total_supply: initial_supply }
+///         }
+///
+///         /// Returns the total supply of the ERC-20 smart contract.
+///         #[ink(message)]
+///         fn total_supply(&self) -> Balance {
+///             self.total_supply
+///         }
+///
+///         // etc ..
+///     }
+/// }
+/// ```
 #[proc_macro_attribute]
 pub fn trait_definition(attr: TokenStream, item: TokenStream) -> TokenStream {
     trait_def::analyze(attr.into(), item.into()).into()
 }
 
-#[cfg(test)]
-pub use contract::generate_or_err;
-
+/// Defines a unit test that makes use of ink!'s off-chain testing capabilities.
+///
+/// If your unit test does not require the existence of an off-chain environment
+/// it is fine to not use this macro since it bears some overhead with the test.
+///
+/// Note that this macro is not required to run unit tests that require ink!'s
+/// off-chain testing capabilities but merely improves code readability.
+///
+/// ## How do you find out if your test requires the off-chain environment?
```

Comment:
I'm not sure if we overcomplicate everything by making the distinction between `[ink::test]` and `#[test]`.

It comes down to a trade-off: performance vs. ease-of-use. I tend towards ease-of-use since the overhead should not be that big if people just use `[ink::test]` everywhere (haven't measured it though). We could get rid of this entire distinction above then.

What do you think?

Reply:
I totally get your point. I think it is just the purist in me trying to make people really understand the difference. If we don't tell them in the first place I fear that confusion might be bigger.